Expanding on the naming conventions (#793)
UniMath/agda-unimath: NAMING-CONVENTIONS.md

Reviewer: I wonder if it would be a good idea to also provide a table of commonly used descriptors. Some that come to mind are
- `compute`
- `htpy`
- `extensionality`
- `is-property`

Perhaps that section is also a good occasion to comment on the use of `'` for variants of the same thing, like when we have a covariant and a contravariant definition, or one that takes implicit and one that takes explicit arguments. Sometimes, I also think it is used when the thing without the `'` fits more naturally as a construction in the given namespace, but the definition with the `'` has weaker assumptions.

Reviewer: Another one is `coherence`

Author: Yep, we can definitely write something about them

Author: I just added a list of common descriptors to the file. I omitted `htpy` because it is directly tied to homotopies, but I could be convinced to put it in there. With such lists, it probably doesn't hurt to put in too much information. I will resolve this conversation here, but have a look if more common descriptors could come up in the list. We can start a new conversation near the source code of the list.

Reviewer: Excellent! I think this helps contextualize the discussion on the general scheme, and gets readers started on forming their own names.

reveal common patterns in names, which hopefully helps to find predictable
namings for entries. On the other hand, we understand that this is not always
possible. If you find yourself not knowing what to name something, give it your
best shot to come up with a name, or reach out to us on the
[Univalent Agda discord server](https://discord.gg/Zp2e8hYsuX) to ask if we have
any suggestions. One of the most commonly asked questions is what to name a
certain entry.

We also mention that the naming scheme of agda-unimath evolves as the library
grows. This implies that there are necessarily some inconsistencies in the
naming of our entries, even though we continually work to improve them.
Sometimes we find that old namings don't fit our ideas of a good naming scheme
anymore, or sometimes we gain a better understanding of what an entry is about
and update its name accordingly. We should therefore remember that neither the
naming scheme nor the names we use in the library are set in stone. If you find
a particular entry where the naming seems off, for any reason, we would love to
know about it. Maintaining and improving it is part of the work of maintaining
the agda-unimath library.

> ## Examples
> 
> For example, the library has an entry named `is-iso-hom-Ring` for the predicate

Reviewer: The transition here is a little abrupt. How about

```suggestion
## Examples

Before we derive a general scheme, let us first get a feel for the structure of names in agda-unimath by considering some examples.
The library has an entry named `is-iso-hom-Ring` for the predicate
```

that a ring homomorphism is an isomorphsim. The most significant aspect of this
predicate is the assertion that something is an isomorphism. Furthermore, we
make this assertion about ring homomorphisms. The name `is-iso-hom-Ring` is
therefore a logical name for the predicate that a ring homomorphism is an
isomorphism.

In our naming scheme we strive for a direct correspondence between a
construction's name and its type. Take, for example, the proof that the
successor function on integers is an equivalence. It has the type
`is-equiv succ-ℤ`, so we name it `is-equiv-succ-ℤ`.

We also consider the type of hypotheses when naming a construction. Typically,
when including hypotheses in the name, we mention them after the type of the
main construction. Let's take the entry `is-equiv-is-contr-map` as an example.
In this entry, we show that any
[contractible map](foundation.contractible-maps.md) is an

> [equivalence](foundation.equivalences.md). The type of this entry is therefore

Reviewer: I tend to link to core, as that is usually where the relevant definition is usually located

```suggestion
[equivalence](foundation-core.equivalences.md). The type of this entry is therefore
```

Author: The reason I usually don't link to core is because except in some special cases I don't remember (and don't want to rememeber) if it has a base file in core. But you're right, linking to core is linking to the actual definitions, and I agree with you.

`(a + b) + (c + d) ＝ (a + c) + (b + d)`. This law is called
`interchange-law-add-add-ℤ`. Again, this naming follows the pattern
`[output-type]-[Namespace]`, because it states the type of which an element is
constructed.

### Further naming conventions for operations on types

It is also common that we have to record computation laws for functions. For
instance,
[transport along identifications](foundation.transport-along-identifications.md)
in a family of [loop spaces](synthetic-homotopy-theory.loop-spaces.md) acts by
[conjugation](synthetic-homotopy-theory.conjugation-loops.md). The name for
transport along identifications is `tr` and the name for loop spaces in the
library is `Ω`. Therefore, we record the function that transport computes to as
`tr-Ω` and we record the [homotopy](foundation.homotopies.md) that transport is
pointwise equal to `tr-Ω` as `compute-tr-Ω`.

## Overview of our naming conventions

- Names are unique; we steer clear of namespace overloading.

- Names should accurately convey the concept of its construction.

- We use US English spelling of words in names.

- When an entry is predominantly about objects of an important concept, the

>   names of these concepts can be capitalized. Usually, these are categories like
>   `Prop`, `Set`, `Semigroup`, `Monoid`, `Group`, `Preorder`, `Poset`,
>   `Precategory`, `Category`, `Directed-Graph`, `Undirected-Graph`, and so on.

Reviewer: I think it is better if the meaning of category is loosened here.
```suggestion
  names of these concepts can be capitalized. Usually, these are categories in one way or another, like
  `Prop`, `Set`, `Semigroup`, `Monoid`, `Group`, `Preorder`, `Poset`,
  `Precategory`, `Category`, `Directed-Graph`, `Undirected-Graph`, `Truncated-Type`, `Spectrum` and so on.
```
and maybe some more exotic example to pique

We also consider the type of hypotheses when naming a construction. Typically,
when including hypotheses in the name, we mention them after the type of the
main construction. Let's take the entry `is-equiv-is-contr-map` as an example.
In this entry, we show that any
[contractible map](foundation.contractible-maps.md) is an
[equivalence](foundation-core.equivalences.md). The type of this entry is
therefore `is-contr-map f → is-equiv f`, where `f` is an assumed function. In
the term `is-equiv-is-contr-map H`, the descriptor `is-contr-map` is positioned
adjacent to its corresponding variable, `H`. Furthermore, by beginning the name
with the descriptor `is-equiv` we quickly see that this entry outputs proofs of
equivalence.

By aligning names with types and incorporating hypotheses when relevant, the
naming scheme aims for clarity and predictability.

## The general scheme

In general, our naming scheme follows the pattern:

```text
  [descriptor]-[output-type]-[input-types]-[Namespace]
```

In this general pattern, all the components are optional. However, their order
is fixed and should be maintained for consistency.

The general naming pattern breaks down as follows:

> - **[descriptor]:** This part is used to give a custom descriptive name for the
>   entry.

Reviewer: ```suggestion
- **[descriptor]:** This part is used to give a custom descriptive name for the
  entry, or refer to the custom descriptive name of another entry.
```
If I didn't misunderstand, this is how it is used in the below example `is-iso-prop-hom-Ring`.

We saw, for example, that the prediate `is-iso-hom-Ring` has the part `Ring`
capitalized. This signifies that the predicate is about rings. This name follows
the scheme `[descriptor]-[input-types]-[Namespace]`. Note that there is also the
entry `is-iso-prop-hom-Ring`. This is a predicate that returns for each ring
homomorphism the _proposition_ that it is an isomorphism, and therefore it
follows the scheme `[descriptor]-[output-type]-[input-types]-[Namespace]`. Now
we can guess what a construction named `hom-iso-Ring` should be about: It should
be a construction that constructs the underlying homomorphism of an isomorphisms
of rings. This name follows the pattern
`[output-type]-[input-types]-[Namespace]`.

There is also a common class of entries where we don't use the `[descriptor]`
part in the name of an entry: underlying objects. For example, the underlying
set of a group is called `set-Group`, which uses the pattern
`[output-type]-[Namespace]`. The construction of the underlying set of a group
returns for each group a set, which is an element of type `Set`. Similarly, we
have `type-Group`, `semigroup-Group`, `type-Ring`, `set-Ring`, and so on.

> Another instance where this happens is in `hom-iso-Group`, which is the

Reviewer: Isn't `hom-iso-Group` an instance of `[output-type]-[input-types]-[Namespace]`?

construction that returns the underlying group homomorphism of an isomorphism of
group. The fact that a group isomorphism is an isomorphsim is called

> `is-iso-iso-Group`, which also uses the pattern `[output-type]-[Namespace]`. One

Reviewer: Isn't this also an instance of `[output-type]-[input-types]-[Namespace]`?

could also consider calling it `is-iso-hom-iso-Group`, to emphasize that the
underlying group homomorphism of the isomorphism is an isomorphism. However,

> this name does not fit our patterns in any way, and the addition of `hom` to the

Reviewer: It's not clear to me how this does not fit our patterns better than the proposed one. Also, the name would be more consistent with `is-iso-hom-inv-iso-Group`

> name adds no extra useful information. This situation is common in instances

Reviewer: It seems to me the added `hom` is disambiguative and does add extra information. To take an example where I more explicitly disagree with this convention, consider your recent change renaming `is-unital-semigroup-Monoid` to `is-unital-Monoid`, but there, the added `semigroup` _did_ add extra useful information. It disambiguated from for instance `is-unital-magma-Monoid` and `is-unital-h-space-Monoid`. Of course, you could argue that semigroup is the correct assumption when the information is omitted, but a reader may not immediately know to conclude that. They shouldn't have to know which way a monoid is defined, as a unital magma that is associative, or a semigroup that is unital.

Reviewer: Also, let me paraphrase from a discussion yesterday:

We use `[property]-[underlying-thing]-[thing]` for very common things in the library like sets and propositions. E.g. `is-set-type-Set`. For the case of `Contr`, the name `is-contr-Contr` is already taken. This begs the question: how do we name properties of the types of objects when they can be interpreted as both properties of the type itself, but also as a property of its objects?

Author: I think your argument that the reader shouldn't have to know what is asserted to be unital is very nice. Our naming scheme should be somewhat intuitive, even for readers who are not well-versed with our naming scheme. I will add something like that to the text. 

On the other hand, the term `is-unital-Monoid` *is* intuitively clear? What does it assert to be unital: the monoid! It extracts in a direct way as possible the unitality of the monoid. That is what the name conveys, and I think the short name conveys this better than `is-unital-semigroup-Monoid`.

I could also see where adding the underlying object could be *more* confusing. For example, `is-category-Category` extracts in a direct way as possible that a category is a category. On the other hand, we could write `is-category-precategory-Category`. This to me reads more like the entry proving that the precategory of categories is a category. 

Your point about the important exceptions, where the disambiguation is necessary, also brings up an important thing that needs to be clarified. If a disambiguation is necessary, for whatever reason, then we disambiguate. But generally we don't adjust our entire naming scheme for that. We work with the nicest possible scheme, that gives concise descriptive names that are well aligned with standard mathematical practice, and if there are exceptions we make exceptions. *Even if the exceptions are in very important cases.*

It reminds me a bit of the early choice I had to make about whether a concepts-first approach to library-design was possible. _Of course_ there were going to be exceptions. Right at the heart of type theory: Identity types, equivalences, dependent pair types, propositions, sets, and so on all had to be split up because it was not strictly possible to maintain a concepts-first design free of cyclic dependencies. But I found it very important to _not let the library design be dictated by a few exceptions_, however prominent those exceptions were. If I would want a super consistent library design, it coudn't be concepts-first and it would probably look more like the standard library.

So, your point about the important exceptions of `Set` and `Contr` is well taken, but we have to carefully decide what to do with those exceptions. Both options are valid choices, but neither are going to be free of ambiguity. And since the longer names are not necessarily more clarifying, I would still have a preference for shorter names, which will work well most of the time, but sometimes need to be disambiguated.

Author: (The disadvantage of the standard library design is of course that you have to know how something is formalized (i.e., as a instance of what general construction or concept) in order to find it. So it may be more consistent, but this consistency comes at the price that it straining the user.

Author: And another point is that if we choose `is-unital-semigroup-Monoid`, then we force the user to know that the monoids are formalised as semigroups in order to use its unitality. If we would like to not require this of the user, then `is-unital-Monoid` would be a better name, because it gives the user access to the unitality of the monoid without them having to know what the name of its underlying structure is.

Reviewer: Ah, haha, I see I failed to observe how the statements coincide. My apologies, then I am on board.

Reviewer: Thank you for spelling this out for me, it is now much clearer where the principles are coming from.

where we omit the `[descriptor]` part of a name. For instance
`is-category-Category` and `is-ideal-ideal-Ring` follow the patterns

> `[output-type]-[Namespace]` and `[output-type]-[input-types]-[Namespace]`,

Reviewer 2: On second reading I'm not sure why
- `is-category-Category : (C : Category) -> is-category C` and
- `is-ideal-ideal-Ring : (I : ideal-Ring) -> is-ideal I`
are different? Even the modules where they are defined seem to follow the exact same structure.

`is-contr-map f → is-equiv f`, where `f` is an assumed function. In the term
`is-equiv-is-contr-map H`, the descriptor `is-contr-map` is positioned adjacent
to its corresponding variable, `H`. Furthermore, by beginning the name with the
descriptor `is-equiv` we quickly see that this entry outputs proofs of
equivalence.

By aligning names with types and incorporating hypotheses when relevant, the
naming scheme aims for clarity and predictability.

## The general scheme

In general, our naming scheme follows the pattern:

```text
  [name]-[type]-[hypotheses]-[Namespace]
```

In this general pattern, all the components are optional. However, their order
is fixed and should be maintained for consistency.

The general naming pattern breaks down as follows:

- **[name]:** This part is used to give a custom descriptive name for the entry.
- **[type]:** This part of the name refers to the output type of the entry.
- **[hypotheses]:** This part of the name consists of descriptors for the
  hypotheses used in the type specification of the entry.
- **[Namespace]:** This part of the name describes what important concept or
  general category the entry is about.

> Given Agda's current lack of a namespace mechanism, we've incorporated what

Author: Agda has a module system. New entries cannot be added to a module once it is closed. A namespace mechanism would allow one to add things to a namespace across several files.

> Given Agda's current lack of a namespace mechanism, we've incorporated what

Reviewer 2: Alright, that makes sense, although it wasn't immediately clear to me - could you maybe rephrase it as "Given Agda's current lack of a namespace mechanism, we can't have named logical collection of definitions that span multiple files. Instead, we've incorporated [..]", or something like that?

### Naming mathematical structures

We saw, for example, that the prediate `is-iso-hom-Ring` has the part `Ring`
capitalized. This signifies that the predicate is about rings. This name follows
the scheme `[name]-[hypotheses]-[Namespace]`. Note that there is also the entry
`is-iso-prop-hom-Ring`. This is a predicate that returns for each ring
homomorphism the _proposition_ that it is an isomorphism, and therefore it
follows the scheme `[name]-[type]-[hypotheses]-[Namespace]`. Now we can guess
what a construction named `hom-iso-Ring` should be about: It should be a
construction that constructs the underlying homomorphism of an isomorphisms of
rings. This name follows the pattern `[type]-[hypotheses]-[Namespace]`.

There is also a common class of entries where we don't use the `[name]` part in
the name of an entry: underlying objects. For example, the underlying set of a
group is called `set-Group`, which uses the pattern `[type]-[Namespace]`. The
construction of the underlying set of a group returns for each group a set,
which is an element of type `Set`. Similarly, we have `type-Group]`,
`semigroup-Group`, `type-Ring`, `set-Ring`, and so on. Another instance where

> this happens is in `hom-iso-Group`, which is the construction that returns the

Reviewer 2: `hom-iso-Group` has the approximate type `iso -> hom`, shouldn't it then be `[type]-[hypotheses]-[Namespace]`?

Author: That looks right

underlying group homomorphism of an isomorphism of group. The fact that a group
isomorphism is an isomorphsim is called `is-iso-iso-Group`, which also uses the

> pattern `[type]-[Namespace]`. One could also consider calling it

Author: Things that go under [type] are the **output types** of the specification of an entry. In the case of `is-iso-hom-Ring` the output type of its specification is `UU`. We don't put this in the name. Instead `is-iso` is a descriptor of the specification, falling under `[name]` in the general pattern.

> pattern `[type]-[Namespace]`. One could also consider calling it

Reviewer 2: I don't think I follow. What I'm reading is that
- `is-iso-iso-Group : (f : iso) -> is-iso f` follows `[type]-[Namespace]`, but
- `is-ideal-ideal-Ring : (I : ideal) -> is-ideal I` follows `[type]-[hypotheses]-[Namespace]`

Is that correct?

`is-iso-hom-iso-Group`, to emphasize that the underlying group homomorphism of
the isomorphism is an isomorphism. However, this name does not fit our patterns
in any way, and the addition of `hom` to the name adds no extra useful
information. This situation is common in instances where we omit the `[name]`
part of a name. For instance `is-category-Category` and `is-ideal-ideal-Ring`
follow the patterns `[type]-[Namespace]` and `[type]-[hypotheses]-[Namespace]`.

Another class of entries where the naming scheme needs some explanation, is
where we define a structured object from another structured object. For
instance, the [kernel](group-theory.kernels.md) of a
[group homomorphism](group-theory.homomorphisms-groups.md) is defined to be the
[normal subgroup](group-theory.normal-subgroups.md) `kernel-hom-Group`. This
name follows the scheme `[name]-[hypotheses]-[Namespace]`. When we want to
define the underlying structure of the kernel of a group homomorphism, we follow

> the scheme `[type]-[hypotheses]-[Namespace]`. For instance, the underlying group

Author: Good point!

of the kernel of a group homomorphism is called `group-kernel-hom-Group`.

### Naming conventions for mathematical laws

Often, mathematical laws are recorded by first specifying in full generality
what it means to satisfy that law. For example, the
[unit laws](foundation.unital-binary-operations.md) for a binary operation are
stated as `left-unit-law` and `right-unit-law`. The fact that
[multiplication on the integers](elementary-number-theory.multiplication-integers.md)
satisfies the unit laws is then stated as `left-unit-law-mul-ℤ` and
`right-unit-law-mul-ℤ`. In the general naming scheme, this naming follows the

> pattern `[type]-[Namespace]`, because it states the type of which an element is

Reviewer 2: I'd also say that this is `[type]-[name]`, or possibly `[type]-[name]-[Namespace]`, because otherwise aren't you saying that `left-unit-law-mul` is a type?

Reviewer 2: Then I don't see where the `[Namespace]` part comes in? Since the `-Z` suffix is already in the `[type]` part.

Author: But if it were counted in the type part, then we'd have to say `interchange-law-add-ℤ-add-ℤ`. Instead, here the proposal is to consider `ℤ` in the namespace part. The name of the type we're landing in of course also mentions the namespace, but in a namespace you don't repeat the name of the namespace for every entry in it that you use. Instead, _within the namespace_ the type would be `left-unit-law mul`, which would appear in the name of an entry as `left-unit-law-mul`, and we append the namespace.

Author: BTW, all of these are very good questions. I will try to clear it up in the text too.

Reviewer 2: Oh, that does make it way clearer, thanks!

constructed, and we treat `ℤ` as a namespace.

For a second example, `interchange-law` states what it means to satisfy the
[interchange law](foundation.interchange-law.md). This interchange law requires
two operations. When we want to prove an interchange law for two specific
operations `mul1` and `mul2`, we name it `interchange-law-mul1-mul2`. We use
this naming scheme, even if the two operations are the same. In fact two
_associative_ operations that satisfy the interchange law will necessarily be
the same. For example, in the [integers](elementary-number-theory.integers.md)
there is an interchange law for addition over addition, which states that
`(a + b) + (c + d) ＝ (a + c) + (b + d)`. This law is called
`interchange-law-add-add-ℤ`. Again, this naming follows the pattern

> `[type]-[Namespace]`, because it states the type of which an element is

Reviewer 2: Again I'd say this is `[type]-[name]-[Namespace]`

Author: The type of the output of this entry is `interchange-law add-ℤ add-ℤ`, so the name here merely describes the type, and the namespace for the integers. In the case `is-equiv succ-ℤ` this would also fall under `[type]-[Namespace]`.

Reviewer 2: The explanation on `left-unit-law-mul-ℤ` makes this clear now 👌

  `Prop`, `Set`, `Semigroup`, `Monoid`, `Group`, `Preorder`, `Poset`,
  `Precategory`, `Category`, `Directed-Graph`, `Undirected-Graph`, and so on.

- As a general rule of thumb, names should start out with an all lowercase
  portion with words separated by hyphens, and may have a capitalized portion at
  the end that describes which larger mathematical framework the definition
  takes place in -- for instance, if it is constructed internally to a certain
  subuniverse or category of mathematical objects.

- The start of a name describes the object that is being constructed. For some
  theorems, the latter part of a name describes the hypotheses.

- We use Unicode symbols sparingly and only when they align with established
  mathematical practice.

- Just as we do with abbreviations, we use special symbols sparingly in names.

- If a symbol is not available, we describe the concept concisely in words.

- We prioritize the readability of the code and avoid subtly different
  variations of the same symbol. An important exception is the use of the
  [full width equals sign](https://codepoints.net/U+ff1d) for the identity type,

>   as the standard equals sign is a reserved symbol in Agda.

Reviewer 2: I'd also mention the homotopy composition and whiskering situation

Reviewer: Other examples are the asterisk vs asterisk operator, and big vee vs small vee. Maybe rephrase to "try to avoid subtly different"

If you are unsure about what to name your entry, ask us on the Univalent Agda

> discord server. One of the most commonly asked questions is what to name a

Reviewer 2: Is this advice intentionally mentioned twice?

Author: Probably it's an oversight. Thanks!

certain thing.

## Naming conventions we try to avoid

- Abbreviations might seem like a good way to shorten names, but we use them
  sparingly. They might save a couple of keystrokes for the author, but in the
  grand scheme of things, they will likely compromise readability and
  maintainability, especially for newcomers and maintainers. We aim for clarity,
  not brevity.

- Using Unicode characters in names is permissible, but we recommend restraint
  to maintain readability. Just a few well-placed symbols can often express a
  lot. Often, when we introduce mixfix operators, we also introduce full names
  for them. We then use the full names for these operators in the naming scheme,
  and omit the symbolic notation for anything but the operation itself. For
  example, the commutativity of cartesian products is called `commutative-prod`
  and not `commutative-×`, and the unit of a modality is called `unit-modality`
  and not `unit-○`.

- To enhance conceptual clarity, we suggest names of constructions avoid
  referring to variable names. This makes code more understandable, even at a
  glance, and easier to work with in subsequent code.

> - Using the projection functions `pr1` and `pr2`, particularly their

Reviewer 2: I'd keep this paragraph where it was, I don't really see how it pertains to naming things - it tells you to add some definitions, not necessarily how to name them

  compositions, can lead to short code, but we recommend to avoid doing so. When
  constructions contain a lot of projections throughout their definition, the
  projections reveal little of what is going on in that part of the projections.
  We therefore prefer naming the projections. When a type of the form `Σ A B` is
  given a name, naming its projections too can enhance readability and will
  provide more informative responses when jumping to the definition.
  Furthermore, it makes it easier to change the definition later on.

- Lastly, we recommend not naming constructions after infix notation of
  operations included in them. Preferring primary prefix notation over infix
  notation can help keep our code consistent. For example, it's preferred to use
  `commutative-prod` instead of `commutative-×` for denoting the commutativity
  of cartesian products.

> - Names never reference variables.

Reviewer 2: This is a restatement of the two paragraphs above

_associative_ operations that satisfy the interchange law will necessarily be
the same. For example, in the [integers](elementary-number-theory.integers.md)
there is an interchange law for addition over addition, which states that
`(a + b) + (c + d) ＝ (a + c) + (b + d)`. This law is called
`interchange-law-add-add-ℤ`. Again, this naming follows the pattern
`[output-type]-[Namespace]`, because it states the type of which an element is
constructed.

### Further naming conventions for operations on types

It is also common that we have to record computation laws for functions. For
instance,
[transport along identifications](foundation.transport-along-identifications.md)
in a family of [loop spaces](synthetic-homotopy-theory.loop-spaces.md) acts by
[conjugation](synthetic-homotopy-theory.conjugation-loops.md). The name for
transport along identifications is `tr` and the name for loop spaces in the
library is `Ω`. Therefore, we record the function that transport computes to as
`tr-Ω` and we record the [homotopy](foundation.homotopies.md) that transport is
pointwise equal to `tr-Ω` as `compute-tr-Ω`.

## List of common descriptors

> In the table below we list some common descriptors that are not directly tied to
> a type with that name.

Reviewer: ```suggestion
To give a sense of the kind of general descriptors we use, we list in the table below some common descriptors that are not directly tied to
a type with that name.
```

Author: Can you edit your suggestion with a period directly after "some common descriptors"? I think it is perhaps better to not restrict ourselves by adding that last part of the sentence.

| Descriptor       | Purpose                                                                                                        |
| ---------------- | -------------------------------------------------------------------------------------------------------------- |
| `coh`            | Used for proofs of coherence                                                                                   |
| `coherence`      | Used to assert coherence                                                                                       |
| `compute`        | Used for proofs of computation, which may be an identification or an element of any other identity system      |

> | `Eq`             | Used for identity systems on types of a particular form                                                        |

Author: I wonder if we shouldn't use `identity-system` where we currently use `Eq` (but I will not make such a change in the current PR)

Reviewer: Yeah, I thought `Eq` was deprecated. I've been using `htpy` for this so far 😬

Author: It's not deprecated. For many types it makes sense to use `htpy` to say that the inhabitants are a structured tuple of homotopies. Similarly, for many types it makes sense to use `equiv` for the descriptor of its identity system. But for some types neither is possible, and then we go with `Eq`. Many number systems are of this kind, and they are certainly not deprecated:)

Author: Actually, `identity-system-X` should return an element of the type of identity systems of `X`, while `Eq` returns the underlying family of types. There is a difference.

Reviewer: I see I've just been making assumptions about the general case from my special cases. 😅 In either case, the capitalized `Eq` is a bit of an eye-sore.

Author: I agree

the same. For example, in the [integers](elementary-number-theory.integers.md)
there is an interchange law for addition over addition, which states that
`(a + b) + (c + d) ＝ (a + c) + (b + d)`. This law is called
`interchange-law-add-add-ℤ`. Again, this naming follows the pattern
`[output-type]-[Namespace]`, because it states the type of which an element is
constructed.

### Further naming conventions for operations on types

It is also common that we have to record computation laws for functions. For
instance,
[transport along identifications](foundation.transport-along-identifications.md)
in a family of [loop spaces](synthetic-homotopy-theory.loop-spaces.md) acts by
[conjugation](synthetic-homotopy-theory.conjugation-loops.md). The name for
transport along identifications is `tr` and the name for loop spaces in the
library is `Ω`. Therefore, we record the function that transport computes to as
`tr-Ω` and we record the [homotopy](foundation.homotopies.md) that transport is
pointwise equal to `tr-Ω` as `compute-tr-Ω`.

## List of common descriptors

To give a sense of the kind of general descriptors we use, we list some common
descriptors in the table below.

> | Descriptor       | Purpose                                                                                                                                                                                           |

Reviewer: Add `statement`

> | Descriptor       | Purpose                                                                                                                                                                                           |

Reviewer: I think it would be very good to have an example column for this table as well. That's what I did for the sHoTT project

> | Descriptor       | Purpose                                                                                                                                                                                           |

Reviewer: Another descriptor pair we use a lot is `is`/`has` :)

Author: Why do you think of them as descriptors? That seems grammatically very weird.

Reviewer: They are modifiers signifying that the type is a property/structure, and then we omit them for the associated total types. It just seems consistent with the scheme.

Author: I'm willing to come up with our own naming convention, but I'm not willing to come up with a new theory that the verbs to be and to have are modifiers. That idea seems out of this world to me.

| `Eq`             | Used for identity systems on types where `htpy` or `equiv` is not an appropriate descriptor of the identity system. `Eq` is commonly used as a descriptor for identity systems on number systems. |
| `eq`             | Used as a descriptor for the identity type                                                                                                                                                        |
| `equiv`          | Used for equivalences, and also for names of identity systems of universe-like types                                                                                                              |
| `extensionality` | Used for computations of identity types                                                                                                                                                           |
| `htpy`           | Used for constructions of homotopies, and also for names of identity systems of function-like types                                                                                               |

> | `is-property`    | Used when `is-prop` is unavailable                                                                                                                                                                |

Reviewer: I don't remember if I already mentioned this, but did you know we actually define `is-property` in `foundation-core.subtypes`? Although IIRC it is not referred to anywhere.

In general, our naming scheme follows the pattern:

```text
  [descriptor]-[output-type]-[input-types]-[Namespace]
```

In this general pattern, all the components are optional. However, their order
is fixed and should be maintained for consistency.

The general naming pattern breaks down as follows:

- **[descriptor]:** This part is used to give a custom descriptive name for the
  entry.
- **[output-type]:** This part of the name refers to the output type of the
  entry.
- **[input-types]:** This part of the name consists of references to the input
  types used in the type specification of the entry.
- **[Namespace]:** This part of the name describes what important concept or
  general category the entry is about.

Given Agda's current lack of a namespace mechanism, we can't have named logical

> collection of definitions that span multiple files. Instead, we've incorporated

Reviewer 2: ```suggestion
collections of definitions that span multiple files. Instead, we've incorporated
```
Sorry I had a typo in my original suggestion.

Author: Perfect, I was going to ask you to use github suggestions. Since you are helping out a lot with this PR, you should be a coauthor after it is merged, and using github's suggestion system is one way to do that.

Reviewer 2: I was using my phone to write the other comments, and it's a lot less convenient to add GH suggestions from the app 😅

what would typically be a namespace directly into the name. This convention is
particularly applied to key mathematical concepts, such as groups, rings,

> categories, graphs, and trees, and so on.

Reviewer 2: ```suggestion
categories, graphs, trees, and so on.
```
